### PR TITLE
feat: validation service

### PR DESCRIPTION
**Pull request by aljo242**
Spins up and internal validation service that can introspect the live prices the sidecar is producing. 

If any of the prices is below our configured liveness threshold (99%, perhaps this should be stricter), the process will terminate itself. 

If all prices are valid after a configured amount of time, the process will gracefully tear down and end sucessfully.

### CODE REVIEW

File: pkg/http/address.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package http
+
+import "net"
+
+func IsValidAddress(address string) bool {
```
Comment (aljo242) [inline pkg/http/address.go:5]: moved to a common folder so it could be used by other code

File: contrib/compose/docker-compose-dev.yml [hunk 2]
```diff
@@ -44,6 +44,7 @@ services:
       "--market-config-path", "/data/markets.json",
       "--pprof-port", "6060",
       "--run-pprof",
+      # "-m", "validate" uncomment to run in validation mode with default config
```
Comment (aljo242) [inline contrib/compose/docker-compose-dev.yml]: how to run in validation mode

File: cmd/constants/marketmaps/markets.go [hunk 3]
```diff
@@ -6354,67 +6259,6 @@ var (
 			  }
 			]
 		  },
-		  "RNDR/USD": {
```
Comment (aljo242) [inline cmd/constants/marketmaps/markets.go:6357]: removed the dead feeds.  Validation fails when they are here and passing when they are gone

File: cmd/connect/main.go [hunk 4]
```diff
@@ -187,6 +200,20 @@ func init() {
...
+		"m",
+		string(modeExec),
+		"Select the mode to run the oracle in.  Default is \"exec\" which will fetch prices as configured.  \"validate\" mode will run the oracle for a set period of time to validate the configuration.",
+	)
+	rootCmd.Flags().IntVarP(
+		&validationPeriod,
+		flagValidationPeriod,
+		"",
+		int(validation.DefaultValidationPeriod),
+		"Duration to run in validation mode.  Note: this flag is only used if mode == \"validate\"",
```
Comment (Eric-Warehime) [inline cmd/connect/main.go]: It's not clear to me what the units are for this int--hours minutes days etc.

File: contrib/compose/docker-compose-dev.yml [hunk 5]
```diff
@@ -44,6 +44,7 @@ services:
       "--market-config-path", "/data/markets.json",
       "--pprof-port", "6060",
       "--run-pprof",
+      "-m", "validate" # uncomment to run in validation mode with default config
```
Comment (Eric-Warehime) [inline contrib/compose/docker-compose-dev.yml]: Is this commented?

File: service/validation/validation.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+			if len(missingPrices) > 0 {
+				v.logger.Warn("currently missing prices", zap.Any("prices", missingPrices))
+				for _, price := range missingPrices {
+					missingPricesMap[price]++
+				}
+			}
+		}
+	}()
+
+	time.Sleep(v.cfg.ValidationPeriod)
```
Comment (Eric-Warehime) [inline service/validation/validation.go]: Can you make this a select loop that checks the context periodically? I feel like graceful shutdown would be nice here especially if our validation period is long.

File: cmd/connect/main.go (hunk 4 above)
Comment (technicallyty) [inline cmd/connect/main.go]: there is `rootCmd.Flags().DurationVarP()`

**Review by Eric-Warehime — APPROVED** [review]

File: cmd/connect/main.go [hunk 7]
```diff
@@ -75,12 +80,21 @@ var (
...
+	validationPeriod    time.Duration
 )
 
 const (
 	DefaultLegacyConfigPath = "./oracle.json"
 )
 
+type runMode string
+
+var (
```
Comment (technicallyty) [inline cmd/connect/main.go]:
```suggestion
const (
```

File: cmd/connect/main.go [hunk 8]
```diff
@@ -187,6 +201,19 @@ func init() {
...
+		flagMode,
+		"m",
+		string(modeExec),
+		"Select the mode to run the oracle in.  Default is \"exec\" which will fetch prices as configured.  \"validate\" mode will run the oracle for a set period of time to validate the configuration.",
+	)
+	rootCmd.Flags().DurationVar(
+		&validationPeriod,
+		flagValidationPeriod,
+		validation.DefaultValidationPeriod,
+		"Duration to run in validation mode in seconds.  Note: this flag is only used if mode == \"validate\"",
```
Comment (technicallyty) [inline cmd/connect/main.go]: with the `DurationVar` change, 1h, 5m, etc is valid now :-)
```suggestion
		"Duration to run in validation mode.  Note: this flag is only used if mode == \"validate\"",
```

File: cmd/connect/main.go [hunk 9]
```diff
@@ -395,6 +424,27 @@ func runOracle() error {
...
+
+		go func(c context.CancelFunc) {
+			defer c()
+
+			_, err := validatorService.Run(ctx)
+			if err != nil {
+				logger.Error("failed to validate metrics", zap.Error(err))
+
+				// kill the process
+				os.Exit(1)
```
Comment (technicallyty) [inline cmd/connect/main.go:441]: careful with this if you still want the `defer c()` to run. deferred functions are not run when using os.Exit. not really sure if that changes any code paths or not

File: contrib/compose/docker-compose-dev.yml (hunk 5 above)
Comment (technicallyty) [inline contrib/compose/docker-compose-dev.yml]: should this be commented out? comment suggests so

File: service/validation/validation.go [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+	"go.uber.org/zap"
+
+	oraclemetrics "github.com/skip-mev/connect/v2/oracle/metrics"
+)
+
+const (
+	DefaultNumChecks             = 1000
+	DefaultRequiredPriceLiveness = 99.0
+	DefaultValidationPeriod      = 10 * time.Minute
+	DefaultBurnInPeriod          = 60 * time.Second
```
Comment (technicallyty) [inline service/validation/validation.go]: the rest of these vars are pretty self explanatory, however  i think `BurnInPeriod` could use a comment

File: service/validation/validation.go [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+	return Validator{
+		logger:  logger.Named("validation"),
+		metrics: metrics,
+		cfg:     cfg,
+	}
+}
+
+// Config includes information for configuring a validation service instance.
+type Config struct {
+	BurnInPeriod          time.Duration
```
Comment (technicallyty) [inline service/validation/validation.go]: or can comment here, either one works.

File: service/validation/validation.go [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+// Run runs the validation service, checking for missing prices and accumulating liveness data.
+func (v *Validator) Run(ctx context.Context) (LivenessResults, error) {
+	v.logger.Info("running in validation mode", zap.Duration("validation period (s)",
+		v.cfg.ValidationPeriod))
+
+	v.logger.Info("waiting for burn in period to end", zap.Duration("period (s)", v.cfg.BurnInPeriod))
+	time.Sleep(v.cfg.BurnInPeriod)
+
+	missingPricesMap := make(map[string]int)
+	tickTime := v.cfg.ValidationPeriod / time.Duration(v.cfg.NumChecks)
```
Comment (technicallyty) [inline service/validation/validation.go:92]: after reading the config, i would have expected the ticker to tick on the duration i passed into `ValidationPeriod`. maybe im just not understanding the code here though.

Comment (technicallyty) [inline service/validation/validation.go:92]: should we validate `tickTime` being a good value? a `0` tickTime will panic

File: service/validation/validation.go
Comment (technicallyty) [inline service/validation/validation.go:1]: should we add a test for this logic?

File: cmd/connect/main.go (hunk 9 above)
Comment (aljo242) [reply cmd/connect/main.go:441]: This means the mm is invalid, so we just want to kill everything immediately.  Will never happen in prod running

**Pull request merged**